text/3371-cargo-target-directories.md

```diff
@@ -0,0 +1,328 @@
+work from its own position to configure the target directory.
+
+On the other hand `CARGO_TARGET_DIRECTORIES` is intended to be used with several projects, possibly completely unrelated to each other. As such,
+it does not accept relative paths, only absolute ones. If a compelling use case is present for a relative path, it can added in the future as a
+backward-compatible change.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+-->
+
+## Setting `CARGO_TARGET_DIRECTORIES`
+
+The option is similar to `CARGO_TARGET_DIR` and can be set in the same places. From less to most specific:
+
+- Through the `config.toml`:
+
+  ```toml
+  [build]
+  target-directories = "/absolute/path/to/target/directories"
+  ```
+
+- Through the environment variable: `CARGO_TARGET_DIRECTORIES="/absolute/path/to/target/directories" cargo build`
+- Through the command line flag: `cargo build --target-directories /absolute/path/to/target/directories`
+
+The given path must be absolute: setting `CARGO_TARGET_DIRECTORIES` to an empty or relative path is an error (when used and not instantly overriden by `CARGO_TARGET_DIR`).
+
+## Resolution order relative to `CARGO_TARGET_DIR`
+
+The resolution order favors `CARGO_TARGET_DIR` in all its forms, in the interest of both backward compatibility and allowing overriding for a singular workspace:
+
+`--target-dir` > `CARGO_TARGET_DIR` > `target-dir = ...` > `--target-directories` > `CARGO_TARGET_DIRECTORIES` > `target-directories = ...`
+
+## Naming
+
+In the example in the previous section, using `CARGO_TARGET_DIRECTORIES` with `cargo build` produces named subdirectories. The name of those is partially deterministic:
+it is the name of the parent directory of the workspace's `Cargo.toml` manifest and an unspecified hash of the absolute path to the workspace's root, so building `work-project/crate-1` will still use the `/cargo-caches/work-project-<hash>/debug/...` directory for a `cargo build` call.
+
+This naming scheme is chosen to be simple for people to navigate but is **not considered stable**: the hashing method (and so the hash) will not change within a single minor version of cargo (1.68.0 -> 1.68.1) but it can change between any two minor versions (1.68 -> 1.69) and tools using that needs to interact with `cargo`'s target directory should not rely on its value for more than a single invocation of them: they should instead query `cargo metadata` for the actual value.
```

Comment:
Please provide the motivation for the stated stability guarantee.

Also, should this be a guarantee or a best-effort policy?  There is a part of me that says "but what if there was a bug or security vulnerability..."

Reply:
I wrote too strong a guarantee, fixed